pageserver/src/tenant/timeline.rs

```diff
@@ -118,6 +121,80 @@ impl PartialOrd for Hole {
     }
 }
 
+pub struct LayerMapping(HashMap<PersistentLayerKey, Arc<dyn PersistentLayer>>);
+
+impl LayerMapping {
+    fn get_from_desc(&self, desc: &PersistentLayerDesc) -> Arc<dyn PersistentLayer> {
+        self.0.get(&desc.key()).expect("not found").clone()
+    }
+
+    pub(crate) fn insert(&mut self, layer: Arc<dyn PersistentLayer>) {
+        self.0.insert(layer.layer_desc().key(), layer);
```

Comment:
I don't think the original code before all refactors start guarantee this, so I put a panic that only happens in debug mode here, hopefully helping us detect errors in tests but not panic in the production env.

Reply:
Assert or no assert is related to #3869 and subsequent revert in #4104 and their follow-ups. I am unsure if it should be dealt here, but instead kept in mind...

Reply:
yeah we found out the hard way that duplicates are possible.

so, we can't just `insert` (and implicitly replace) here.

Reply:
If there are duplicates, they will have the same file name on disk, and on next restart there should be exactly one entry for that key? Therefore, implicit replace should behave the same as before, I suppose…?

Reply:
> If there are duplicates, they will have the same file name on disk, and on next restart there should be exactly one entry for that key? Therefore, implicit replace should behave the same as before, I suppose…?

Hm, we encountered duplicates after #3869 and hence reverted the assert in #4104.
Quoting from the revert:

> As noted in PR https://github.com/neondatabase/neon/pull/4094, we do in fact try to insert duplicates to the layer map, if L0->L1 compaction is interrupted. We do not have a proper fix for that right now, and we are in a hurry to make a release to production, so revert the changes related to this to the state that we have in production currently. We know that we have a bug here, but better to live with the bug that we've had in production for a long time, than rush a fix to production without testing it in staging first.

Then you merged

* #4408 

which switched to implicit replacement, IIUC.
So, we have the implicit replacement of duplicates, identified by matching key & lsn range, in prod already.
So far, nobody complained.
Doesn't mean it's correct though.

Someone with deeper understanding about why duplicates were expected to be absent should re-review #4408 cc @knizhnik @koivunej 
I'll call you responsible for leading that investigation :) 

Regardless, the implicit replace is not technically a blocker for this PR then, because it's pre-existing since #4408 . Just bear in mind that, if we need to fix that aspect of #4408 , it'll need to ripple through this PR stack.

Reply:
> Then you merged #4408 which switched to implicit replacement

could @problame link to a place where the semantics were changed in #4408? I think what we had before (by my recollection after #4104) was that inserts just replaced silently, and I don't see #4408 changing anything. Or do I misremember the in-between behaviour?

Reply:
I think I understand that correctly. Resolving this.